File: impulse-macos/Sources/ImpulseApp/Sidebar/FileIcons.swift
import AppKit

// MARK: - Icon Color Field

/// Which theme color an icon uses, mirroring the Rust `ColorField` enum.
enum IconColorField {
    case orange, blue, yellow, green, cyan, magenta, red, comment, fg

    func resolve(_ theme: Theme) -> String {
        switch self {
        case .orange:  return theme.orangeHex
        case .blue:    return theme.blueHex
        case .yellow:  return theme.yellowHex
        case .green:   return theme.greenHex
        case .cyan:    return theme.cyanHex
        case .magenta: return theme.magentaHex
        case .red:     return theme.redHex
        case .comment: return theme.commentHex
        case .fg:      return theme.fgHex
        }
    }
}

// MARK: - Icon Definition

private struct IconDef {
    let name: String
    let color: IconColorField
}

/// All icon definitions, matching `file_icons.rs` exactly.
private let allIcons: [IconDef] = [
    // Languages
    IconDef(name: "rust", color: .orange),
    IconDef(name: "python", color: .blue),
    IconDef(name: "javascript", color: .yellow),
    IconDef(name: "typescript", color: .blue),
    IconDef(name: "go", color: .cyan),
    IconDef(name: "c", color: .blue),
    IconDef(name: "cpp", color: .blue),
    IconDef(name: "java", color: .orange),
    IconDef(name: "kotlin", color: .magenta),
    IconDef(name: "swift", color: .orange),
    IconDef(name: "ruby", color: .red),
    IconDef(name: "php", color: .magenta),
    IconDef(name: "csharp", color: .green),
    IconDef(name: "zig", color: .orange),
    IconDef(name: "haskell", color: .magenta),
    IconDef(name: "lua", color: .blue),
    IconDef(name: "dart", color: .cyan),
    IconDef(name: "elixir", color: .magenta),
    IconDef(name: "scala", color: .magenta),
    IconDef(name: "clojure", color: .green),
    IconDef(name: "erlang", color: .red),
    IconDef(name: "nim", color: .yellow),
    IconDef(name: "julia", color: .magenta),
    IconDef(name: "r", color: .blue),
    IconDef(name: "tex", color: .orange),
    // Web
    IconDef(name: "html", color: .orange),
    IconDef(name: "css", color: .blue),
    IconDef(name: "sass", color: .magenta),
    IconDef(name: "vue", color: .green),
    IconDef(name: "svelte", color: .orange),
    IconDef(name: "react", color: .cyan),
    // Data / Config
    IconDef(name: "json", color: .yellow),
    IconDef(name: "yaml", color: .orange),
    IconDef(name: "toml", color: .orange),
    IconDef(name: "xml", color: .comment),
    IconDef(name: "markdown", color: .blue),
    IconDef(name: "settings", color: .comment),
    // Tooling
    IconDef(name: "console", color: .green),
    IconDef(name: "docker", color: .cyan),
    IconDef(name: "git", color: .orange),
    IconDef(name: "lock", color: .comment),
    IconDef(name: "database", color: .cyan),
    // Media
    IconDef(name: "image", color: .green),
    IconDef(name: "audio", color: .magenta),
    IconDef(name: "video", color: .magenta),
    IconDef(name: "pdf", color: .red),
    // General
    IconDef(name: "document", color: .fg),
    IconDef(name: "archive", color: .comment),
    IconDef(name: "binary", color: .comment),
    // Folders
    IconDef(name: "folder", color: .cyan),
    IconDef(name: "folder-open", color: .cyan),
    // Toolbar
    IconDef(name: "toolbar-sidebar", color: .fg),
    IconDef(name: "toolbar-plus", color: .fg),
    IconDef(name: "toolbar-eye-open", color: .fg),
    IconDef(name: "toolbar-eye-closed", color: .fg),
    IconDef(name: "toolbar-collapse", color: .fg),
    IconDef(name: "toolbar-refresh", color: .fg),
]

// MARK: - Icon Name Lookup

/// Maps a filename to its icon name, exactly mirroring the Rust `lookup_icon_name`.
func lookupIconName(filename: String, isDirectory: Bool, expanded: Bool) -> String {
    if isDirectory {
        return expanded ? "folder-open" : "folder"
    }

    let ext = filename.split(separator: ".").last.map(String.init)?.lowercased() ?? ""
    switch ext {
    // Languages
    case "rs": return "rust"
    case "py", "pyi", "pyw": return "python"
    case "js", "mjs", "cjs": return "javascript"
    case "ts", "mts", "cts": return "typescript"
    case "go": return "go"
    case "c", "h": return "c"
    case "cpp", "cc", "cxx", "hpp", "hxx", "hh": return "cpp"
    case "java": return "java"
    case "kt", "kts": return "kotlin"
    case "swift": return "swift"
    case "rb", "erb": return "ruby"
    case "php": return "php"
    case "cs": return "csharp"
    case "zig": return "zig"
    case "hs", "lhs": return "haskell"
    case "lua": return "lua"
    case "dart": return "dart"
    case "ex", "exs", "heex": return "elixir"
    case "scala", "sc": return "scala"
    case "clj", "cljs", "cljc", "edn": return "clojure"
    case "erl", "hrl": return "erlang"
    case "nim", "nims": return "nim"
    case "jl": return "julia"
    case "r", "rmd": return "r"
    case "tex", "sty", "cls", "bib": return "tex"
    // Web
    case "html", "htm": return "html"
    case "css": return "css"
    case "scss", "sass", "less": return "sass"
    case "vue": return "vue"
    case "svelte": return "svelte"
    case "jsx", "tsx": return "react"
    // Data / Config
    case "json", "jsonc", "json5": return "json"
    case "yaml", "yml": return "yaml"
    case "toml": return "toml"
    case "xml", "xsl", "xslt", "xsd", "wsdl": return "xml"
    case "md", "mdx", "markdown": return "markdown"
    case "ini", "cfg", "conf", "ron", "properties": return "settings"
    // Shell / Tooling
    case "sh", "bash", "zsh", "fish", "ps1", "bat", "cmd": return "console"
    case "lock": return "lock"
    case "sql", "sqlite", "db": return "database"
    // Media
    case "png", "jpg", "jpeg", "gif", "svg", "ico", "webp", "bmp", "tiff", "avif": return "image"
    case "mp3", "wav", "flac", "ogg", "aac", "wma", "m4a": return "audio"
    case "mp4", "mkv", "avi", "webm", "mov", "wmv", "flv": return "video"
    case "pdf": return "pdf"
    // Archives
    case "zip", "tar", "gz", "bz2", "xz", "7z", "rar", "zst", "tgz": return "archive"
    // Binary / Executables
    case "exe", "dll", "so", "dylib", "a", "o", "wasm": return "binary"
    default: return lookupByFilename(filename)
    }
}

/// Fallback lookup by full filename for special files.
private func lookupByFilename(_ filename: String) -> String {
    switch filename.lowercased() {
    case "dockerfile", "containerfile": return "docker"
    case "makefile", "rakefile", "justfile", "taskfile": return "console"
    case ".gitignore", ".gitmodules", ".gitattributes": return "git"
    case "license", "licence", "license.md", "licence.md", "license.txt", "licence.txt": return "document"
    case "readme", "readme.md", "readme.txt": return "document"
    case "changelog", "changelog.md", "authors", "contributing", "contributing.md": return "document"
    case "cargo.toml", "cargo.lock": return "rust"
    case "package.json", "package-lock.json": return "javascript"
    case "tsconfig.json": return "typescript"
    case "go.mod", "go.sum": return "go"
    case "gemfile", "gemfile.lock": return "ruby"
    case "composer.json", "composer.lock": return "php"
    case ".eslintrc", ".prettierrc", ".editorconfig": return "settings"
    default: return "document"
    }
}

// MARK: - SVG Recoloring

/// Replaces hex fill/stroke attribute values in an SVG string with a theme color.
private func recolorSVG(_ svg: String, color: String) -> String {
    let colorValue = color.hasPrefix("#") ? color : "#\(color)"
    var result = svg

    // Replace fill="#XXXXXX" and stroke="#XXXXXX" with the theme color
    if let regex = try? NSRegularExpression(pattern: ##"(fill|stroke)="#[0-9A-Fa-f]{3,8}""##) {
        result = regex.stringByReplacingMatches(
            in: result,
            range: NSRange(result.startIndex..., in: result),
            withTemplate: "$1=\"\(colorValue)\""
        )
    }
    return result
}

// MARK: - Icon Cache

/// Pre-renders all SVG icons for the current theme and caches the resulting NSImages.
final class IconCache {
    private var images: [String: NSImage] = [:]
    private var svgStrings: [String: String] = [:]
    private var iconDefs: [String: IconDef] = [:]

    init(theme: Theme) {
        // Index icon defs by name
        for def in allIcons {
            iconDefs[def.name] = def
        }
        loadSVGs()
        build(theme: theme)
    }

    /// Loads raw SVG strings from the bundle resource directory.
    private func loadSVGs() {
        guard let iconsURL = Bundle.module.url(forResource: "icons", withExtension: nil) else {
            NSLog("IconCache: icons resource directory not found in bundle")
            return
        }

        for def in allIcons {
            let svgURL = iconsURL.appendingPathComponent("\(def.name).svg")
            if let svgString = try? String(contentsOf: svgURL, encoding: .utf8) {
                svgStrings[def.name] = svgString
            }
        }
    }

    /// Rebuilds the cache for a new theme.
    func rebuild(theme: Theme) {
        images.removeAll()
        build(theme: theme)
    }

    private func build(theme: Theme) {
        for def in allIcons {
            guard let svg = svgStrings[def.name] else { continue }
            let color = def.color.resolve(theme)
            let recolored = recolorSVG(svg, color: color)
            if let image = renderSVG(recolored, size: 16) {
                images[def.name] = image
            }
        }
    }

    /// Renders an SVG string to an NSImage at the given point size.
    private func renderSVG(_ svg: String, size: CGFloat) -> NSImage? {
        guard let data = svg.data(using: .utf8) else { return nil }
        guard let image = NSImage(data: data) else { return nil }
        image.size = NSSize(width: size, height: size)
        return image
    }

    /// Returns the themed icon for a file or directory.
    func icon(filename: String, isDirectory: Bool, expanded: Bool) -> NSImage? {
        let name = lookupIconName(filename: filename, isDirectory: isDirectory, expanded: expanded)
        return images[name]
    }

    /// Returns a toolbar icon by name (e.g. "toolbar-sidebar", "console", "settings").
    func toolbarIcon(name: String) -> NSImage? {
        return images[name]
    }
}

File: impulse-linux/src/file_icons.rs
use std::collections::HashMap;

use gtk4::gdk;
use gtk4::glib;

use crate::theme::ThemeColors;

// ---------------------------------------------------------------------------
// Embedded SVGs (Material Icon Theme, MIT license)
// ---------------------------------------------------------------------------

// Languages
const RUST_SVG: &str = include_str!("../../assets/icons/rust.svg");
const PYTHON_SVG: &str = include_str!("../../assets/icons/python.svg");
const JAVASCRIPT_SVG: &str = include_str!("../../assets/icons/javascript.svg");
const TYPESCRIPT_SVG: &str = include_str!("../../assets/icons/typescript.svg");
const GO_SVG: &str = include_str!("../../assets/icons/go.svg");
const C_SVG: &str = include_str!("../../assets/icons/c.svg");
const CPP_SVG: &str = include_str!("../../assets/icons/cpp.svg");
const JAVA_SVG: &str = include_str!("../../assets/icons/java.svg");
const KOTLIN_SVG: &str = include_str!("../../assets/icons/kotlin.svg");
const SWIFT_SVG: &str = include_str!("../../assets/icons/swift.svg");
const RUBY_SVG: &str = include_str!("../../assets/icons/ruby.svg");
const PHP_SVG: &str = include_str!("../../assets/icons/php.svg");
const CSHARP_SVG: &str = include_str!("../../assets/icons/csharp.svg");
const ZIG_SVG: &str = include_str!("../../assets/icons/zig.svg");
const HASKELL_SVG: &str = include_str!("../../assets/icons/haskell.svg");
const LUA_SVG: &str = include_str!("../../assets/icons/lua.svg");
const DART_SVG: &str = include_str!("../../assets/icons/dart.svg");
const ELIXIR_SVG: &str = include_str!("../../assets/icons/elixir.svg");
const SCALA_SVG: &str = include_str!("../../assets/icons/scala.svg");
const CLOJURE_SVG: &str = include_str!("../../assets/icons/clojure.svg");
const ERLANG_SVG: &str = include_str!("../../assets/icons/erlang.svg");
const NIM_SVG: &str = include_str!("../../assets/icons/nim.svg");
const JULIA_SVG: &str = include_str!("../../assets/icons/julia.svg");
const R_SVG: &str = include_str!("../../assets/icons/r.svg");
const TEX_SVG: &str = include_str!("../../assets/icons/tex.svg");

// Web
const HTML_SVG: &str = include_str!("../../assets/icons/html.svg");
const CSS_SVG: &str = include_str!("../../assets/icons/css.svg");
const SASS_SVG: &str = include_str!("../../assets/icons/sass.svg");
const VUE_SVG: &str = include_str!("../../assets/icons/vue.svg");
const SVELTE_SVG: &str = include_str!("../../assets/icons/svelte.svg");
const REACT_SVG: &str = include_str!("../../assets/icons/react.svg");

// Data / Config
const JSON_SVG: &str = include_str!("../../assets/icons/json.svg");
const YAML_SVG: &str = include_str!("../../assets/icons/yaml.svg");
const TOML_SVG: &str = include_str!("../../assets/icons/toml.svg");
const XML_SVG: &str = include_str!("../../assets/icons/xml.svg");
const MARKDOWN_SVG: &str = include_str!("../../assets/icons/markdown.svg");
const SETTINGS_SVG: &str = include_str!("../../assets/icons/settings.svg");

// Tooling
const CONSOLE_SVG: &str = include_str!("../../assets/icons/console.svg");
const DOCKER_SVG: &str = include_str!("../../assets/icons/docker.svg");
const GIT_SVG: &str = include_str!("../../assets/icons/git.svg");
const LOCK_SVG: &str = include_str!("../../assets/icons/lock.svg");
const DATABASE_SVG: &str = include_str!("../../assets/icons/database.svg");

// Media
const IMAGE_SVG: &str = include_str!("../../assets/icons/image.svg");
const AUDIO_SVG: &str = include_str!("../../assets/icons/audio.svg");
const VIDEO_SVG: &str = include_str!("../../assets/icons/video.svg");
const PDF_SVG: &str = include_str!("../../assets/icons/pdf.svg");

// General
const DOCUMENT_SVG: &str = include_str!("../../assets/icons/document.svg");
const ARCHIVE_SVG: &str = include_str!("../../assets/icons/archive.svg");
const BINARY_SVG: &str = include_str!("../../assets/icons/binary.svg");

// Folders
const FOLDER_SVG: &str = include_str!("../../assets/icons/folder.svg");
const FOLDER_OPEN_SVG: &str = include_str!("../../assets/icons/folder-open.svg");

// Toolbar
const TOOLBAR_SIDEBAR_SVG: &str = include_str!("../../assets/icons/toolbar-sidebar.svg");
const TOOLBAR_PLUS_SVG: &str = include_str!("../../assets/icons/toolbar-plus.svg");
const TOOLBAR_EYE_OPEN_SVG: &str = include_str!("../../assets/icons/toolbar-eye-open.svg");
const TOOLBAR_EYE_CLOSED_SVG: &str = include_str!("../../assets/icons/toolbar-eye-closed.svg");
const TOOLBAR_COLLAPSE_SVG: &str = include_str!("../../assets/icons/toolbar-collapse.svg");
const TOOLBAR_REFRESH_SVG: &str = include_str!("../../assets/icons/toolbar-refresh.svg");
const TOOLBAR_NEW_FILE_SVG: &str = include_str!("../../assets/icons/toolbar-new-file.svg");
const TOOLBAR_NEW_FOLDER_SVG: &str = include_str!("../../assets/icons/toolbar-new-folder.svg");
const PIN_SVG: &str = include_str!("../../assets/icons/pin.svg");

// ---------------------------------------------------------------------------
// Color field — which theme color an icon uses
// ---------------------------------------------------------------------------

#[derive(Clone, Copy)]
enum ColorField {
    Orange,
    Blue,
    Yellow,
    Green,
    Cyan,
    Magenta,
    Red,
    Comment,
    Fg,
}

impl ColorField {
    fn resolve(self, theme: &ThemeColors) -> &'static str {
        match self {
            ColorField::Orange => theme.orange,
            ColorField::Blue => theme.blue,
            ColorField::Yellow => theme.yellow,
            ColorField::Green => theme.green,
            ColorField::Cyan => theme.cyan,
            ColorField::Magenta => theme.magenta,
            ColorField::Red => theme.red,
            ColorField::Comment => theme.comment,
            ColorField::Fg => theme.fg,
        }
    }
}

// ---------------------------------------------------------------------------
// Icon registry — each icon has a name, SVG template, and color field
// ---------------------------------------------------------------------------

struct IconDef {
    name: &'static str,
    svg: &'static str,
    color: ColorField,
}

const ALL_ICONS: &[IconDef] = &[
    // Languages
    IconDef {
        name: "rust",
        svg: RUST_SVG,
        color: ColorField::Orange,
    },
    IconDef {
        name: "python",
        svg: PYTHON_SVG,
        color: ColorField::Blue,
    },
    IconDef {
        name: "javascript",
        svg: JAVASCRIPT_SVG,
        color: ColorField::Yellow,
    },
    IconDef {
        name: "typescript",
        svg: TYPESCRIPT_SVG,
        color: ColorField::Blue,
    },
    IconDef {
        name: "go",
        svg: GO_SVG,
        color: ColorField::Cyan,
    },
    IconDef {
        name: "c",
        svg: C_SVG,
        color: ColorField::Blue,
    },
    IconDef {
        name: "cpp",
        svg: CPP_SVG,
        color: ColorField::Blue,
    },
    IconDef {
        name: "java",
        svg: JAVA_SVG,
        color: ColorField::Orange,
    },
    IconDef {
        name: "kotlin",
        svg: KOTLIN_SVG,
        color: ColorField::Magenta,
    },
    IconDef {
        name: "swift",
        svg: SWIFT_SVG,
        color: ColorField::Orange,
    },
    IconDef {
        name: "ruby",
        svg: RUBY_SVG,
        color: ColorField::Red,
    },
    IconDef {
        name: "php",
        svg: PHP_SVG,
        color: ColorField::Magenta,
    },
    IconDef {
        name: "csharp",
        svg: CSHARP_SVG,
        color: ColorField::Green,
    },
    IconDef {
        name: "zig",
        svg: ZIG_SVG,
        color: ColorField::Orange,
    },
    IconDef {
        name: "haskell",
        svg: HASKELL_SVG,
        color: ColorField::Magenta,
    },
    IconDef {
        name: "lua",
        svg: LUA_SVG,
        color: ColorField::Blue,
    },
    IconDef {
        name: "dart",
        svg: DART_SVG,
        color: ColorField::Cyan,
    },
    IconDef {
        name: "elixir",
        svg: ELIXIR_SVG,
        color: ColorField::Magenta,
    },
    IconDef {
        name: "scala",
        svg: SCALA_SVG,
        color: ColorField::Magenta,
    },
    IconDef {
        name: "clojure",
        svg: CLOJURE_SVG,
        color: ColorField::Green,
    },
    IconDef {
        name: "erlang",
        svg: ERLANG_SVG,
        color: ColorField::Red,
    },
    IconDef {
        name: "nim",
        svg: NIM_SVG,
        color: ColorField::Yellow,
    },
    IconDef {
        name: "julia",
        svg: JULIA_SVG,
        color: ColorField::Magenta,
    },
    IconDef {
        name: "r",
        svg: R_SVG,
        color: ColorField::Blue,
    },
    IconDef {
        name: "tex",
        svg: TEX_SVG,
        color: ColorField::Orange,
    },
    // Web
    IconDef {
        name: "html",
        svg: HTML_SVG,
        color: ColorField::Orange,
    },
    IconDef {
        name: "css",
        svg: CSS_SVG,
        color: ColorField::Blue,
    },
    IconDef {
        name: "sass",
        svg: SASS_SVG,
        color: ColorField::Magenta,
    },
    IconDef {
        name: "vue",
        svg: VUE_SVG,
        color: ColorField::Green,
    },
    IconDef {
        name: "svelte",
        svg: SVELTE_SVG,
        color: ColorField::Orange,
    },
    IconDef {
        name: "react",
        svg: REACT_SVG,
        color: ColorField::Cyan,
    },
    // Data / Config
    IconDef {
        name: "json",
        svg: JSON_SVG,
        color: ColorField::Yellow,
    },
    IconDef {
        name: "yaml",
        svg: YAML_SVG,
        color: ColorField::Orange,
    },
    IconDef {
        name: "toml",
        svg: TOML_SVG,
        color: ColorField::Orange,
    },
    IconDef {
        name: "xml",
        svg: XML_SVG,
        color: ColorField::Comment,
    },
    IconDef {
        name: "markdown",
        svg: MARKDOWN_SVG,
        color: ColorField::Blue,
    },
    IconDef {
        name: "settings",
        svg: SETTINGS_SVG,
        color: ColorField::Comment,
    },
    // Tooling
    IconDef {
        name: "console",
        svg: CONSOLE_SVG,
        color: ColorField::Green,
    },
    IconDef {
        name: "docker",
        svg: DOCKER_SVG,
        color: ColorField::Cyan,
    },
    IconDef {
        name: "git",
        svg: GIT_SVG,
        color: ColorField::Orange,
    },
    IconDef {
        name: "lock",
        svg: LOCK_SVG,
        color: ColorField::Comment,
    },
    IconDef {
        name: "database",
        svg: DATABASE_SVG,
        color: ColorField::Cyan,
    },
    // Media
    IconDef {
        name: "image",
        svg: IMAGE_SVG,
        color: ColorField::Green,
    },
    IconDef {
        name: "audio",
        svg: AUDIO_SVG,
        color: ColorField::Magenta,
    },
    IconDef {
        name: "video",
        svg: VIDEO_SVG,
        color: ColorField::Magenta,
    },
    IconDef {
        name: "pdf",
        svg: PDF_SVG,
        color: ColorField::Red,
    },
    // General
    IconDef {
        name: "document",
        svg: DOCUMENT_SVG,
        color: ColorField::Fg,
    },
    IconDef {
        name: "archive",
        svg: ARCHIVE_SVG,
        color: ColorField::Comment,
    },
    IconDef {
        name: "binary",
        svg: BINARY_SVG,
        color: ColorField::Comment,
    },
    // Folders
    IconDef {
        name: "folder",
        svg: FOLDER_SVG,
        color: ColorField::Cyan,
    },
    IconDef {
        name: "folder-open",
        svg: FOLDER_OPEN_SVG,
        color: ColorField::Cyan,
    },
    // Toolbar
    IconDef {
        name: "toolbar-sidebar",
        svg: TOOLBAR_SIDEBAR_SVG,
        color: ColorField::Fg,
    },
    IconDef {
        name: "toolbar-plus",
        svg: TOOLBAR_PLUS_SVG,
        color: ColorField::Fg,
    },
    IconDef {
        name: "toolbar-eye-open",
        svg: TOOLBAR_EYE_OPEN_SVG,
        color: ColorField::Fg,
    },
    IconDef {
        name: "toolbar-eye-closed",
        svg: TOOLBAR_EYE_CLOSED_SVG,
        color: ColorField::Fg,
    },
    IconDef {
        name: "toolbar-collapse",
        svg: TOOLBAR_COLLAPSE_SVG,
        color: ColorField::Fg,
    },
    IconDef {
        name: "toolbar-refresh",
        svg: TOOLBAR_REFRESH_SVG,
        color: ColorField::Fg,
    },
    IconDef {
        name: "toolbar-new-file",
        svg: TOOLBAR_NEW_FILE_SVG,
        color: ColorField::Fg,
    },
    IconDef {
        name: "toolbar-new-folder",
        svg: TOOLBAR_NEW_FOLDER_SVG,
        color: ColorField::Fg,
    },
    IconDef {
        name: "pin",
        svg: PIN_SVG,
        color: ColorField::Comment,
    },
];

// ---------------------------------------------------------------------------
// Extension / filename → icon name lookup
// ---------------------------------------------------------------------------

fn lookup_icon_name(filename: &str, is_dir: bool, expanded: bool) -> &'static str {
    if is_dir {
        return if expanded { "folder-open" } else { "folder" };
    }

    let ext = filename.rsplit('.').next().unwrap_or("");
    match ext.to_lowercase().as_str() {
        // Languages
        "rs" => "rust",
        "py" | "pyi" | "pyw" => "python",
        "js" | "mjs" | "cjs" => "javascript",
        "ts" | "mts" | "cts" => "typescript",
        "go" => "go",
        "c" | "h" => "c",
        "cpp" | "cc" | "cxx" | "hpp" | "hxx" | "hh" => "cpp",
        "java" => "java",
        "kt" | "kts" => "kotlin",
        "swift" => "swift",
        "rb" | "erb" => "ruby",
        "php" => "php",
        "cs" => "csharp",
        "zig" => "zig",
        "hs" | "lhs" => "haskell",
        "lua" => "lua",
        "dart" => "dart",
        "ex" | "exs" | "heex" => "elixir",
        "scala" | "sc" => "scala",
        "clj" | "cljs" | "cljc" | "edn" => "clojure",
        "erl" | "hrl" => "erlang",
        "nim" | "nims" => "nim",
        "jl" => "julia",
        "r" | "rmd" => "r",
        "tex" | "sty" | "cls" | "bib" => "tex",

        // Web
        "html" | "htm" => "html",
        "css" => "css",
        "scss" | "sass" | "less" => "sass",
        "vue" => "vue",
        "svelte" => "svelte",
        "jsx" | "tsx" => "react",

        // Data / Config
        "json" | "jsonc" | "json5" => "json",
        "yaml" | "yml" => "yaml",
        "toml" => "toml",
        "xml" | "xsl" | "xslt" | "xsd" | "wsdl" => "xml",
        "md" | "mdx" | "markdown" => "markdown",
        "ini" | "cfg" | "conf" | "ron" | "properties" => "settings",

        // Shell / Tooling
        "sh" | "bash" | "zsh" | "fish" | "ps1" | "bat" | "cmd" => "console",
        "lock" => "lock",
        "sql" | "sqlite" | "db" => "database",

        // Media
        "png" | "jpg" | "jpeg" | "gif" | "svg" | "ico" | "webp" | "bmp" | "tiff" | "avif" => {
            "image"
        }
        "mp3" | "wav" | "flac" | "ogg" | "aac" | "wma" | "m4a" => "audio",
        "mp4" | "mkv" | "avi" | "webm" | "mov" | "wmv" | "flv" => "video",
        "pdf" => "pdf",

        // Archives
        "zip" | "tar" | "gz" | "bz2" | "xz" | "7z" | "rar" | "zst" | "tgz" => "archive",

        // Binary / Executables
        "exe" | "dll" | "so" | "dylib" | "a" | "o" | "wasm" => "binary",

        // Default: check special filenames
        _ => lookup_by_filename(filename),
    }
}

fn lookup_by_filename(filename: &str) -> &'static str {
    match filename.to_lowercase().as_str() {
        "dockerfile" | "containerfile" => "docker",
        "makefile" | "rakefile" | "justfile" | "taskfile" => "console",
        ".gitignore" | ".gitmodules" | ".gitattributes" => "git",
        "license" | "licence" | "license.md" | "licence.md" | "license.txt" | "licence.txt" => {
            "document"
        }
        "readme" | "readme.md" | "readme.txt" => "document",
        "changelog" | "changelog.md" | "authors" | "contributing" | "contributing.md" => "document",
        "cargo.toml" | "cargo.lock" => "rust",
        "package.json" | "package-lock.json" => "javascript",
        "tsconfig.json" => "typescript",
        "go.mod" | "go.sum" => "go",
        "gemfile" | "gemfile.lock" => "ruby",
        "composer.json" | "composer.lock" => "php",
        ".eslintrc" | ".prettierrc" | ".editorconfig" => "settings",
        _ => "document",
    }
}

// ---------------------------------------------------------------------------
// SVG recoloring — replaces hex fill/stroke values with theme color
// ---------------------------------------------------------------------------

fn replace_attr_colors(svg: &str, attr: &str, color: &str) -> String {
    let pattern = format!("{attr}=\"#");
    let prefix = format!("{attr}=\"");
    let mut result = String::with_capacity(svg.len());
    let mut remaining = svg;

    while let Some(pos) = remaining.find(&pattern) {
        result.push_str(&remaining[..pos]);
        result.push_str(&prefix);
        result.push_str(color);
        result.push('"');
        let value_start = pos + prefix.len();
        if let Some(end_quote) = remaining[value_start..].find('"') {
            remaining = &remaining[value_start + end_quote + 1..];
        } else {
            remaining = "";
            break;
        }
    }
    result.push_str(remaining);
    result
}

fn recolor_svg(svg: &str, color: &str) -> String {
    let result = replace_attr_colors(svg, "fill", color);
    replace_attr_colors(&result, "stroke", color)
}

// ---------------------------------------------------------------------------
// SVG → gdk::Texture rendering via resvg
// ---------------------------------------------------------------------------

const ICON_RENDER_SIZE: u32 = 32;

fn render_svg_to_texture(svg: &str, color: &str) -> Option<gdk::Texture> {
    let recolored = recolor_svg(svg, color);
    let opts = resvg::usvg::Options::default();
    let tree = resvg::usvg::Tree::from_data(recolored.as_bytes(), &opts).ok()?;

    let mut pixmap = resvg::tiny_skia::Pixmap::new(ICON_RENDER_SIZE, ICON_RENDER_SIZE)?;

    // Scale SVG to fit within the target size, centered
    let svg_size = tree.size();
    let sx = ICON_RENDER_SIZE as f32 / svg_size.width();
    let sy = ICON_RENDER_SIZE as f32 / svg_size.height();
    let scale = sx.min(sy);
    let tx = (ICON_RENDER_SIZE as f32 - svg_size.width() * scale) / 2.0;
    let ty = (ICON_RENDER_SIZE as f32 - svg_size.height() * scale) / 2.0;

    let transform = resvg::tiny_skia::Transform::from_scale(scale, scale).post_translate(tx, ty);
    resvg::render(&tree, transform, &mut pixmap.as_mut());

    let png_data = pixmap.encode_png().ok()?;
    let bytes = glib::Bytes::from(&png_data);
    gdk::Texture::from_bytes(&bytes).ok()
}

// ---------------------------------------------------------------------------
// Icon cache — pre-renders all icons for the current theme
// ---------------------------------------------------------------------------

pub struct IconCache {
    textures: HashMap<&'static str, gdk::Texture>,
}

impl IconCache {
    pub fn new(theme: &ThemeColors) -> Self {
        let mut cache = Self {
            textures: HashMap::with_capacity(ALL_ICONS.len()),
        };
        cache.build(theme);
        cache
    }

    pub fn rebuild(&mut self, theme: &ThemeColors) {
        self.textures.clear();
        self.build(theme);
    }

    fn build(&mut self, theme: &ThemeColors) {
        for icon in ALL_ICONS {
            let color = icon.color.resolve(theme);
            if let Some(texture) = render_svg_to_texture(icon.svg, color) {
                self.textures.insert(icon.name, texture);
            }
        }
    }

    pub fn get(&self, filename: &str, is_dir: bool, expanded: bool) -> Option<&gdk::Texture> {
        let name = lookup_icon_name(filename, is_dir, expanded);
        self.textures.get(name)
    }

    /// Returns a toolbar icon texture by name (e.g. "toolbar-sidebar", "console", "settings").
    pub fn get_toolbar_icon(&self, name: &str) -> Option<&gdk::Texture> {
        self.textures.get(name)
    }
}
